src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -1615,6 +1615,52 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
     (t, f2)
   }
 
+  def reorderSamples(newIds: java.util.ArrayList[Annotation]): VariantSampleMatrix[RPK, RK, T] =
+    reorderSamples(newIds.asScala.toArray)
+
+  def reorderSamples(newIds: Array[Annotation]): VariantSampleMatrix[RPK, RK, T] = {
+    requireUniqueSamples("reorder_samples")
+
+    val oldOrder = sampleIds.zipWithIndex.toMap
+    val newOrder = newIds.zipWithIndex.toMap
+
+    val mapping = oldOrder.outerJoin(newOrder).map { case (s, (oldIdx, newIdx)) =>
+      (oldIdx, newIdx) match {
+        case (Some(i), Some(j)) => (i, j)
+        case (None, Some(j)) => fatal(s"`$s' is not in dataset.")
+        case (Some(i), None) => fatal(s"dataset sample id `$s' not found in new id list.")
+        case (None, None) => fatal(s"This case should never happen!")
```

Comment:
maybe just use @unchecked?

Reply:
Also, this will produce an error message that contains one random sample ID that doesn't match. Shouldn't we either produce all of them, or the first?